consensus/src/dag/dag_driver.rs

```diff
@@ -114,3 +131,22 @@ impl DagDriver {
         }
     }
 }
+
+impl RpcHandler for DagDriver {
+    type Request = CertifiedNode;
+    type Response = CertifiedAck;
+
+    fn process(&mut self, node: Self::Request) -> anyhow::Result<Self::Response> {
+        let epoch = node.metadata().epoch();
+        {
+            let dag_reader = self.dag.read();
+            if dag_reader.exists(node.metadata()) {
+                return Ok(CertifiedAck::new(node.metadata().epoch()));
+            }
+        }
+
+        self.add_node(node)?;
```

Comment:
I still don't really convinced that in case we do not have the parents we should rely on the sender to retransmit. Instead of storing it as pending. 
Remind me why we chose this path?

Reply:
We will send the node to dag fetcher, so we are not dropping it.

Reply:
They we should send back Ack, right?

Reply:
One reason is that it might be misleading because its not actually in the DAG (not sure if it matters much) 
Another reason is, in DAG fetcher, we also plan to try to fetch once from all nodes and give up eventually, so we might need retransmission.

Reply:
we should only send back Ack if it's successfully added to the DAG, if the fetch failed we rely on the sender to retransmit for retry